### PR TITLE
osd: add 'ceph tell osd.id smart'

### PR DESCRIPTION
**Pull request by yaarith**
WIP: Adding SMART data collection from OSDs.

### CODE REVIEW

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -6418,6 +6425,37 @@ void OSD::do_command(Connection *con, ceph_tid_t tid, vector<string>& cmd, buffe
...
+    dout(1) << "smartctl is called" << dendl;
+    set<string> devnames;
+    store->get_devices(&devnames);
+    uint64_t smart_timeout = cct->_conf->get_val<uint64_t>("osd_smart_report_timeout");
+    std::string result;
+    
+    json_spirit::mObject json_map; // == typedef std::map<std::string, mValue> mObject;
+    json_spirit::mValue smart_json;
+
+    for (auto it = devnames.begin(); it != devnames.end(); ++it) {
```
Comment (liewegas) [inline src/osd/OSD.cc]: for (auto dev : devnames) {

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -6418,6 +6425,37 @@ void OSD::do_command(Connection *con, ceph_tid_t tid, vector<string>& cmd, buffe
...
+    store->get_devices(&devnames);
+    uint64_t smart_timeout = cct->_conf->get_val<uint64_t>("osd_smart_report_timeout");
+    std::string result;
+    
+    json_spirit::mObject json_map; // == typedef std::map<std::string, mValue> mObject;
+    json_spirit::mValue smart_json;
+
+    for (auto it = devnames.begin(); it != devnames.end(); ++it) { 
+      // smartctl works only on physical devices; filter out any non-physical device
+      if (it->find("dm-") != std::string::npos)
```
Comment (liewegas) [inline src/osd/OSD.cc]: if (dev.find("dm-") == 0) {
  continue;
}

Comment (liewegas) [reply src/osd/OSD.cc]: and you can replace *it with dev through the rest of this too.  the c++11 for loop style is much more concise :)

File: src/osd/OSD.cc [hunk 3]
```diff
@@ -6418,6 +6425,37 @@ void OSD::do_command(Connection *con, ceph_tid_t tid, vector<string>& cmd, buffe
...
+      ss << "smartctl output: " << result; // TODO: remove when json is ready?
+      // TODO: change to read_or_throw?
+      if (!json_spirit::read(result, smart_json)) {
+	derr << "smartctl JSON output of /dev/" + *it + " is invalid" << dendl;
+      }
+      else { //json is valid, assigning 
+	json_map[*it] = smart_json;
+      }
+      // no need to result.clear() or clear smart_json
+    }
```
Comment (liewegas) [inline src/osd/OSD.cc]: here there will eventually be a loop over json_map that assembles the final result, right?

File: src/osd/OSD.cc (hunk 1 above)
Comment (yaarith) [reply src/osd/OSD.cc]: great, done!

File: src/osd/OSD.cc (hunk 2 above)
Comment (yaarith) [reply src/osd/OSD.cc]: sleek, thanks!

just double checking that it should be '!=': 
if (dev.find("dm-") != 0) {
continue;
}

File: src/osd/OSD.cc (hunk 3 above)
Comment (yaarith) [reply src/osd/OSD.cc]: right, I just couldn't test it so far (due to the vm), but will do on mira.

File: src/osd/OSD.cc (hunk 2 above)
Comment (liewegas) [reply src/osd/OSD.cc]: in this case, ==.. find() returns the position, and we're checking of "dm-" is the prefix of the string starting at offset 0

Comment (yaarith) [reply src/osd/OSD.cc]: oops, right, sorry!
is there a possibility that "dm-" itself will have another prefix in the device list?

Comment (liewegas) [reply src/osd/OSD.cc]: i don't think so!

Comment (yaarith) [reply src/osd/OSD.cc]: great, so we got that covered.
still need to take care of RAID devices which may not return SMART data.

File: src/osd/OSD.cc [hunk 4]
```diff
@@ -6457,6 +6491,10 @@ void OSD::do_command(Connection *con, ceph_tid_t tid, vector<string>& cmd, buffe
     ss << "compacted omap in " << duration << " seconds";
   }
 
+  else if (prefix == "smart") {
+    probe_smart(ss);
```
Comment (liewegas) [inline src/osd/OSD.cc]: ds instead of ss (for these calls ss goes to stderr and ds goes to stdout)

Comment (yaarith) [reply src/osd/OSD.cc]: thanks, done.

**Pull request merged**